### ISSUE TITLE
Consider using Gomega and Ginkgo in Cloud Foundry platform tests

### ISSUE THREAD

**Issue by rosenhouse** [istio/istio#2203]
Follow-up to https://github.com/istio/istio/pull/2022

If you have general thoughts about how we should or should-not be standardizing on testing libraries in Istio, please leave your comments in https://github.com/istio/istio/issues/2202

**Comment by rosenhouse** [comment]
@ZackButcher we're trying to understand the rationale here, so that we can make the minimum required changes and not spend too much time re-writing existing tests.

In your objections to Gomega & Ginkgo, we heard a few concerns:
1. future changes are easier for maintainers if there's one style of tests used across the Istio codebase
2. you don't like to grow the set of test imports
3. you don't like dot imports in tests

Here's an issue that we've run into as we look into how we could remove Ginkgo & Gomega:
- our current tests of the `cloudfoundry` package use `code.cloudfoundry.org/copilot/testhelpers` to generate certificates that mimic those served by Copilot, in order to support a mock copilot server
- that package uses Gomega matchers internally, but does not expose them

If we continue to use that `testhelpers` package, then the pilot tests would have a transitive dependency on Gomega, even if we didn't use Gomega directly in the Istio tests.

Would that be ok, or are you really opposed to Gomega existing in the set of transitive dependencies for tests?

An alternative is for us to re-write the functionality from copilot `testhelpers` inside of Istio.  We're not sure that code duplication is worth the effort.

**Comment by ZackButcher** [comment]
I don't want to require any solution that'd force changes to dependencies - that's too much to expect people to do for Istio in general IMO.

I'm not opposed to introducing the dependencies, my main concern w.r.t. maintainability is in the test code that lives inside of the Istio repo itself: I want it to all be as consistent as possible to make it easier for any developer on the project to fix things.

**Comment by kyessenov** [comment]
Retitling the issue. I don't think the test dependency contributes to sprawling build dependency problem.

**Comment by rosenhouse** [comment]
We are about to assemble another PR that will add more features for the Cloud Foundry adapter.

Should we test it using Ginkgo/Gomega, or using Go test?

**Comment by costinm** [comment]
Sorry, I don't have enough context here - and not familiar enough with ginko/gomega vs go tests,
however:
- if there are strong reasons to use them - please configure them with a formatter compatible with go tests and the gotest-to-junit.xml we are now using in CircleCI. 
- it is not good for the project to have each component use its own test (or logging) framework. If we can't
avoid it, or everyone agrees standard go test has problems - make a proposal, document it and we can 
discuss using it across istio.

**Comment by rosenhouse** [comment]
Sorry I forgot to update this issue -- yesterday I opened PR #2345 to remove Ginkgo from the Cloud Foundry unit tests.  So the output should be vanilla gotest.

After #2345 it is merged, we could put together with a subsequent PR that also removes Gomega, or we could leave that as the final state.  Perhaps I will make a proposal in favor of the latter.

**Comment by mandarjog** [comment]
In my view gomega/ginkgo tests are better structured and it is easier to parse intent as compared with the Vanilla table driven go tests.
Kubernetes/tests/e2e uses these for example https://github.com/kubernetes/kubernetes/blob/7d759ba0c9c0af27370a7012f34f372a59123443/test/e2e/kubectl/portforward.go#L460

**Comment by costinm** [comment]
My only concern is consistency - if a logging / testing / grpc library has
clear benefits and broad community support - we should document the reasons
and use it in all modules.

I personally don't like the style - and wasted time already due do output
inconsistencies, but it's just my opinion.
